monai/deploy/operators/nii_data_loader_operator.py

```diff
@@ -0,0 +1,53 @@
+# Copyright 2021-2022 MONAI Consortium
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#     http://www.apache.org/licenses/LICENSE-2.0
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+import SimpleITK
+
+import monai.deploy.core as md
+from monai.deploy.core import DataPath, ExecutionContext, InputContext, IOType, Operator, OutputContext
+
+
+@md.input("image_path", DataPath, IOType.DISK)
+@md.output("image", np.ndarray, IOType.IN_MEMORY)
+class NiftiDataLoader(Operator):
+    """
+    This operator reads a nifti image, extracts the numpy array and forwards it to the next operator
+    """
+
+    def compute(self, op_input: InputContext, op_output: OutputContext, context: ExecutionContext):
+        input_path = op_input.get().path
+        image_np = self.convert_and_save(input_path)
+        op_output.set(image_np)
+
+    def convert_and_save(self, nii_path):
+        """
+        reads the nifti image and
+        """
+        image_reader = SimpleITK.ImageFileReader()
+        image_reader.SetFileName(str(nii_path))
+        image = image_reader.Execute()
+        image_np = np.transpose(SimpleITK.GetArrayFromImage(image), [2, 1, 0])
+        return image_np
```

Comment:
So this will only return the ndarray, how would the metadata be preserved? Metadata, e.g. spacing and affine are important and needed for the inference transforms.

In App SDK there is a Image domain class, very primitive with only ndarray and a list of metadata in a dict. We used this object to preserve image ndarray as well as a list of required metadata from the DICOM Series to vol image. The built-in inference operators are expecting Image type as input.

Also, with NIfTI image only, unless the original slices are converted to a DICOM series, the DICOM Seg Writer operator can not be used, simply because the Seg needs to refer back to the original DICOM instances.

Reply:
Unless the input images in NIfTI have already been resampled to the spacing/orientation to fit the model, resampling and orientation are needed in the pre-processing step, granted in many cases and in competitions the training and validation images have been pre-prepared.
The MONAI built-in readers simply returns the ndarray and a metadata dict containing certain entries common/required by the transforms. The LoadImage class then convert both into a MetaTensor object.